cluster/gce/upgrade.sh

```diff
@@ -343,51 +343,61 @@ function do-node-upgrade() {
-    echo "== Calling rolling-update for ${group}. ==" >&2
-    update=$(gcloud alpha compute rolling-updates \
+    gcloud compute instance-groups managed set-instance-template "${group}" \
+      --template="${template_name}" \
+      --project="${PROJECT}" \
+      --zone="${ZONE}"
+    set_instance_template_rc=$?
+    if [[ "${set_instance_template_rc}" != 0 ]]; then
+      return ${set_instance_template_rc}
+    fi
+    gcloud compute instance-groups managed list-instances "${group}" \
+        --format='value(instance)' \
         --project="${PROJECT}" \
         --zone="${ZONE}" \
-        start \
-        --group="${group}" \
-        --template="${template_name}" \
-        --instance-startup-timeout=300s \
-        --max-num-concurrent-instances=1 \
-        --max-num-failed-instances=0 \
-        --min-instance-update-time=0s 2>&1) && update_rc=$? || update_rc=$?
-
-    if [[ "${update_rc}" != 0 ]]; then
-      echo "== FAILED to start rolling-update: =="
-      echo "${update}"
-      echo "  This may be due to a preexisting rolling-update;"
-      echo "  see https://github.com/kubernetes/kubernetes/issues/33113 for details."
-      echo "  All rolling-updates in project ${PROJECT} zone ${ZONE}:"
-      gcloud alpha compute rolling-updates \
+        | while read instance; do
+      echo "== Draining ${instance}. == " >&2
+      "${KUBE_ROOT}/cluster/kubectl.sh" drain --delete-local-data --force --ignore-daemonsets "${instance}"
+      drain_rc=$?
+      if [[ "${drain_rc}" != 0 ]]; then
+        echo "== FAILED to drain ${instance} =="
+        return ${drain_rc}
+      fi
+      gcloud compute instance-groups managed recreate-instances "${group}" \
         --project="${PROJECT}" \
         --zone="${ZONE}" \
-        list || true
-      return ${update_rc}
-    fi
-
-    id=$(echo "${update}" | grep "Started" | cut -d '/' -f 11 | cut -d ']' -f 1)
-    updates+=("${id}")
-  done
-
-  echo "== Waiting for Upgrading nodes to be finished. ==" >&2
-  # Wait until rolling updates are finished.
-  for update in ${updates[@]}; do
-    while true; do
-      result=$(gcloud alpha compute rolling-updates \
-          --project="${PROJECT}" \
-          --zone="${ZONE}" \
-          describe \
-          ${update} \
-          --format='value(status)' || true)
-      if [ $result = "ROLLED_OUT" ]; then
-        echo "Rolling update ${update} is ${result} state and finished."
-        break
+        --instances="${instance}"
+      recreate_rc=$?
+      if [[ "${recreate_rc}" != 0 ]]; then
+        echo "== FAILED to recreate ${instance} =="
+        return ${recreate_rc}
       fi
-      echo "Rolling update ${update} is still in ${result} state."
-      sleep 10
+      echo "== Waiting for ${instance} to become young. ==" >&2
+      while true; do
+        age=$(( $(date +%s) - $(date --date=$("${KUBE_ROOT}/cluster/kubectl.sh"  get node "${instance}" --output='jsonpath={.metadata.creationTimestamp}' ) +%s) ))
```

Comment:
I agree about the timing flakiness, and I agree I'd like to be sure that the instance has been recreated.

But both of those can happen and it can still be the case that the kubelet hasn't deleted and recreated the node object, so we still need to make sure that the node is `Ready`, and we need to make sure we're checking that the *new* node is healthy, because we may observe the old node with the same name still being `Ready`.

I think I can do this if the script itself deletes the node once it's been drained.  The order would be

1. `kubectl drain`
2. `kubectl delete node`
3. `gcloud ... recreate-instances`
4. poll `gcloud compute operations list`
5. poll `kubectl get node` waiting for it to not have `SchedulingDisabled` and *to* have `Ready`.

This depends on a different assumption about kubelet's behavior, that it won't re-add itself after I delete it.  That seems to be the case if nothing happens on the node, but if kubelet randomly gets restarted (or changes its behavior), it will get re-added, and it will no longer be drained.  IOW, deleting the node opens the door for the node to acquire pods again after the code thinks it's drained.  A different race.

So having written all that, I'm inclined to poll the operation.  You're right that seems cleaner.  But I still need a way to distinguish old node from new node, and I can't think of a better way than polling for age.  I guess I could poll for version with `kubectl get -o jsonpath='{.status.nodeInfo}' node/foo`.  That would work as long as we're sure that we're actually changing kubelet versions.